### ISSUE TITLE
Consistency requirements when using multiple compilers in a DAG

### ISSUE THREAD

**Issue by alalazo** [LLNL/spack#31361]
# Summary

As a user I want Spack to ensure consistency of runtime libraries among different  compilers employed in the same DAG, so that I can use specific compilers only where needed and otherwise use more mainstream compilers.

# Description of the use case

There could be two different variations of this use case. The first one is when different nodes use different compiler suites for the same languages. The second one is when we use different compiler suites for different languages. The two cases can also be happening together in the same DAG.

## Different nodes using different compiler suites

Assume we want to concretize `hdf5` with `clang` and `zlib` with `gcc`:
```console
% spack spec hdf5+cxx~mpi %clang ^zlib %gcc 
```

The resulting DAG might look like:
```mermaid
flowchart TD;
  subgraph one[hdf5 unification space]
  NODE1[hdf5 v1.12.1] --type=link--> NODE2[zlib v1.2.11]
  NODE1 --runtime=c\ntype=link--> NODE3[glibc v2.31]
  NODE2 --runtime=c\ntype=link--> NODE3
  NODE4[libstdc++ v3.4.28] --type=link--> NODE3
  NODE1 --runtime=cxx\ntype=link--> NODE4
  end
  NODE1 --type=build\nlanguage=c\nlanguage=cxx--> NODE5[clang v12.0.1]
  NODE2 --type=build\nlanguage=c--> NODE6[gcc v9.3.0]
  NODE5 -.enforces.-> NODE4
  NODE5 -.enforces.-> NODE3
  NODE6 -.enforces.-> NODE3
  style NODE5 fill:#bbf,stroke:#f66,stroke-width:2px,color:#fff,stroke-dasharray: 5 5
  style NODE6 fill:#bbf,stroke:#f66,stroke-width:2px,color:#fff,stroke-dasharray: 5 5
  style NODE4 fill:#aaa,stroke:#f66,stroke-width:2px,color:#fff,stroke-dasharray: 5 5
  style NODE3 fill:#aaa,stroke:#f66,stroke-width:2px,color:#fff,stroke-dasharray: 5 5
```
Since the C++ runtime can be selected for `clang` we might also decide to use `libc++`:

```console
% spack spec hdf5+cxx~mpi ^libc++ %clang ^zlib %gcc 
```
The resulting DAG would be in this case:
```mermaid
flowchart TD;
  subgraph one[hdf5 unification space]
  NODE1[hdf5 v1.12.1] --type=link--> NODE2[zlib v1.2.11]
  NODE1 --runtime=c\ntype=link--> NODE3[glibc v2.31]
  NODE2 --runtime=c\ntype=link--> NODE3
  NODE4[libc++ v12.0.1] --type=link--> NODE3
  NODE1 --runtime=cxx\ntype=link--> NODE4
  end
  NODE1 --type=build\nlanguage=c\nlanguage=cxx--> NODE5[clang v12.0.1]
  NODE2 --type=build\nlanguage=c--> NODE6[gcc v9.3.0]
  NODE5 -.enforces.-> NODE4
  NODE5 -.enforces.-> NODE3
  NODE6 -.enforces.-> NODE3
  style NODE5 fill:#bbf,stroke:#f66,stroke-width:2px,color:#fff,stroke-dasharray: 5 5
  style NODE6 fill:#bbf,stroke:#f66,stroke-width:2px,color:#fff,stroke-dasharray: 5 5
  style NODE4 fill:#aaa,stroke:#f66,stroke-width:2px,color:#fff,stroke-dasharray: 5 5
  style NODE3 fill:#aaa,stroke:#f66,stroke-width:2px,color:#fff,stroke-dasharray: 5 5
```

## Different languages using different compiler suites

Assume we now want to concretize `hdf5` using `clang` for C and C++ and `gcc` for Fortran:

```console
% spack spec hdf5+cxx+fortran~mpi %:lang=c,cxx: clang %:lang=fortran: gcc 
```

The resulting DAG would look like:
```mermaid
flowchart TD;
  subgraph one[hdf5 unification space]
  NODE1[hdf5 v1.12.1] --type=link--> NODE2[zlib v1.2.11]
  NODE1 --runtime=c\ntype=link--> NODE3[glibc v2.31]
  NODE2 --runtime=c\ntype=link--> NODE3
  NODE4[libstdc++ v3.4.28] --type=link--> NODE3
  NODE1 --runtime=cxx\ntype=link-->NODE4
  NODE1 --runtime=fortran\ntype=link-->NODE7[libgfortran v5]
  NODE7 --type=link-->NODE3
  end
  NODE1 --type=build\nlanguage=c\nlanguage=cxx--> NODE5[clang v12.0.1]
  NODE2 --type=build\nlanguage=c--> NODE5
  NODE1 --type=build\nlanguage=fortran--> NODE6[gcc v9.3.0]
  NODE5 -.enforces.-> NODE4
  NODE5 -.enforces.-> NODE3
  NODE6 -.enforces.-> NODE7
  style NODE5 fill:#bbf,stroke:#f66,stroke-width:2px,color:#fff,stroke-dasharray: 5 5
  style NODE6 fill:#bbf,stroke:#f66,stroke-width:2px,color:#fff,stroke-dasharray: 5 5
  style NODE4 fill:#aaa,stroke:#f66,stroke-width:2px,color:#fff,stroke-dasharray: 5 5
  style NODE3 fill:#aaa,stroke:#f66,stroke-width:2px,color:#fff,stroke-dasharray: 5 5
  style NODE7 fill:#aaa,stroke:#f66,stroke-width:2px,color:#fff,stroke-dasharray: 5 5
```

# Acceptance criteria

1. Define the DAG structure when multiple compilers are used as dependencies
2. Define how to add information on global constraints that might be enforced in the DAG, depending on which compiler and compiler features are used. A use case we know about is OpenMP runtimes when e.g. using `cray`, `intel ` and `gcc` compilers in the same DAG.
3. Define at which granularity we want to model runtime libraries, given that there are multiple configurations possible (e.g. `clang` using `libc++` and `libgcc_s`)

# Additional notes 

Runtime libraries have a certain degree of forward compatibility, and we might want to model that. This is important for packaging binaries, since we can package stuff that requires a minimum GLIBCXX version and deploy it on a system having a later version. *This use case is a practical application of splicing.*

A runtime library may have additional dependencies that can be modeled as well. For instance the C++ runtime provided by `libstdc++` on Ubuntu 20.04 is depending on `libc`, `libm` and `libgcc_s` which can be modeled too. For reference this is what `ldd` outputs:
```
$ ldd /lib/x86_64-linux-gnu/libstdc++.so.6
	linux-vdso.so.1 (0x00007ffef673b000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fa8421d4000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fa841fe2000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fa84251f000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007fa841fc7000)
```

Different compilers need to enforce compatible runtimes in a single unification space

Some compilers can operate with more than one runtime, so choosing a runtime imposes constraints on which compilers can be used in a unification space and vice-versa. These constraints might be global, e.g. they can influence the behavior of nodes which are more than one edge away.[^1]

Using some runtimes might imply injecting flags during compilation e.g. `-stdlib=libc++` see [here](https://libcxx.llvm.org/UsingLibcxx.html)

Using some runtimes might also imply that a compiler is inherently using another compilers (e.g. `clang` using `gcc` for `libstdc++`)

The support for different runtime libraries from a given compiler, e.g. `clang`, is tied to how the compiler was built. This means that properties on the compiler node enforce constraints on the runtime libraries that can be used. Below we see that `libc++.so` from clang can be configure  to use `libgcc_s` from gcc:
```console
$ ldd /home/culpo/opt/compilers/store/linux-ubuntu20.04-icelake/gcc-9.3.0/llvm-12.0.1-orsuvw5rbgn7cxrqxsfkyomgzkm4shxo/lib/libc++.so.1
	linux-vdso.so.1 (0x00007fff3d3f2000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fc94071b000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fc940529000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007fc94050e000)
	libc++abi.so.1 => /home/culpo/opt/compilers/store/linux-ubuntu20.04-icelake/gcc-9.3.0/llvm-12.0.1-orsuvw5rbgn7cxrqxsfkyomgzkm4shxo/lib/libc++abi.so.1 (0x00007fc9404d3000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fc940846000)
```
It is also possible to use `--rtlib=compiler-rt`. That requires to use proper flags and for certain build-systems pass appropriate options (e.g. CMake needs `-DLIBCXX_USE_COMPILER_RT=YES` etc.)

[^1]: One use case we know about is related to OpenMP runtimes. There are cases when, using multiple compilers for different nodes with OpenMP support, we constrain which runtime needs to be used for the entire unification space.

# References
- [ABI Compatibility libstdc++](https://gcc.gnu.org/onlinedocs/libstdc++/manual/abi.html)
- [libc++](https://libcxx.llvm.org) (includes both using system and custom libraries)
- [clang vs. gcc](https://stackoverflow.com/a/63476425/771663) (SO question on whether clang uses GCC)
- [assembling toolchains with LLVM](https://clang.llvm.org/docs/Toolchain.html#linker)

# Related issues
- #18156

**Comment by haampie** [comment]
> Using some runtimes might imply injecting flags during compilation e.g. -stdlib=libc++ see [here](https://libcxx.llvm.org/UsingLibcxx.html)

> Using some runtimes might also imply that a compiler is inherently using another compilers (e.g. clang using gcc for libstdc++)

What's somewhat problematic here is that we don't really have fine-grained control over clang (and its forks...). A system clang for example scans the system for GCC installations:

```
$ clang -v
Found candidate GCC installation: /usr/bin/../lib/gcc/x86_64-linux-gnu/10
Found candidate GCC installation: /usr/bin/../lib/gcc/x86_64-linux-gnu/11
Found candidate GCC installation: /usr/bin/../lib/gcc/x86_64-linux-gnu/7
Found candidate GCC installation: /usr/bin/../lib/gcc/x86_64-linux-gnu/7.5.0
Found candidate GCC installation: /usr/bin/../lib/gcc/x86_64-linux-gnu/8
Found candidate GCC installation: /usr/bin/../lib/gcc/x86_64-linux-gnu/9
Selected GCC installation: /usr/bin/../lib/gcc/x86_64-linux-gnu/11
```

and then picks the latest, from which it takes stuff like `crt[begin|end].o` (for constructors/destructors of globals), "runtime" libraries for gcc itself: `libgcc_s` / `libgcc` and language type of libs `libstdc++`.

It has a `--gcc-toolchain` flag, but it does not help for system GCC's, since it only accepts `/usr` as a prefix, not the actual location it detected above: `--gcc-toolchain=/usr/bin/../lib/gcc/x86_64-linux-gnu/11` fails.

So, basically clang relies on the assumption that the runtime libstdc++ has the newest ABI, and it can't really be influenced to link to a specific libstdc++, except when LLVM itself is installed by Spack (since we fix the GCC prefix at compile time).

Fortunately NVHPC (clang based) can be configured to use a particular GCC. For other LLVM-based secret sauce forks I'm not aware if this is always possible.

---

It may actually be worth submitting a patch to fix clang here...

**Comment by trws** [comment]
It's mildly annoying, but we can actually control for that @haampie. We need to make a directory that symlinks, *sigh*, to the actual gcc path we want from a path that looks like `$PREFIX/include|lib/gcc/$PLATFORM/$VERSION/` to do it, but then `--gcc-toolchain` works.  That's how we maintain clang's with a pre-set default gcc-toolchain in LC for example, the `LIBCXX_GCC_TOOLCHAIN` flag to the cmake needs the same thing.  It's on my list for the llvm package, but that darned directory structure does not make it a nice thing to do, especially for a one-off.
